base/reflection.jl

```diff
@@ -1106,6 +1112,69 @@ function methods(@nospecialize(f),
     return methods(f, Tuple{Vararg{Any}}, mod)
 end
 
+function _instancemethods(t::Type, @nospecialize(mt), lim::Int, world::UInt)
+    tt = instance_signature_type(t, mt)
+    return _methods_by_ftype(tt, lim, world)
+end
+
+"""
+    instancemethods(t::Type, [types], [module])
+
+Return the method list for instances of type `t`.
+
+If `types` is specified, return the methods whose types match.
+If `module` is specified, return the methods defined in that module.
+Multiple modules can also be specified as an array.
+
+# Example
+
+The `instancemethods` function enables the return of methods defined for a callable
+object of a given type, eliminating the need to instantiate the object first.
+
+```julia
+julia> struct A{T}
+    x::T
+end
+
+# making the object callable
+julia> (a::A{Float64})(x::Float64) = a.x + x
+julia> (a::A{String})(x::String) = a.x * x
+
+julia> a = A(1.0)
+A{Float64}(1.0)
+
+julia> methods(a) == instancemethods(A{Float64})
+true
+
+julia> length(instancemethods(A{<:Any})) == 2
+true
+```
+
+!!! compat "Julia 1.4"
+    At least Julia 1.4 is required for specifying a module.
+
+See also: [`methods`](@ref).
+"""
+function instancemethods(t::Type,
+    @nospecialize(tmatch),
+    mod::Union{Tuple{Module},AbstractArray{Module},Nothing}=nothing)
+    world = get_world_counter()
+    ms = Method[]
+    for m in _instancemethods(t, tmatch, -1, world)::Vector
+        m = m::Core.MethodMatch
+        (mod === nothing || parentmodule(m.method) ∈ mod) && push!(ms, m.method)
+    end
```

Comment:
Since this code is nearly identical to that in `methods`, perhaps there's a way to unify them? For example, you could define a helper function like
```julia
function _collectmethods(list::Vector, mod)
    ms = Method[]
    for m in list
        m = m::Core.MethodMatch
        (mod === nothing || parentmodule(m.method) ∈ mod) && push!(ms, m.method)
    end
    return ms
end
```
to use in both `instancemethods` and `methods`, which would pass in the output of `_instancemethods` and `_methods`, respectively, as the list argument.

Reply:
Why not something like
```
function _collectmethods(list, mod)
    [m.method for m in list if mod === nothing || parentmodule(m.method) in mod]
end
```